File: 01matrix.cpp
class Solution {
public:
int m, n;
    vector<vector<int>> updateMatrix(vector<vector<int>>& mat) {
         m=mat.size();
         n=mat[0].size();
        vector<vector<int>> dirs{{0,1},{1,0},{0,-1},{-1,0}};
        queue<pair<int,int>> q;
        for(int i=0;i<m;i++)
        {
            for(int j=0;j<n;j++)
            {
                if(mat[i][j]==0)
                    q.push({i,j});
                else
                    mat[i][j]=-1;
            }
        }

        while(!q.empty())
        {
            auto curr=q.front();
            q.pop();
            for(auto dir:dirs)
            {
                int newR=curr.first+dir[0];
                int newC=curr.second+dir[1];
                if(valid(newR,newC) and mat[newR][newC]==-1)
                {
                    q.push({newR,newC});
                    mat[newR][newC]=mat[curr.first][curr.second]+1;
                }
            }
        }
        return mat;
    }
    bool valid(int r, int c)
    {
        return r>=0 and c>=0 and r<m and c<n;
    }

};

File: _01 Matrix.py
# TC : O(mxn)
# SC : O(mxn)

class Solution(object):
    def updateMatrix(self, mat):
        """
        :type mat: List[List[int]]
        :rtype: List[List[int]]
        """
        q = deque()
        m = len(mat)
        n = len(mat[0])
        dist = 1
        dirs = [[1,0],[0,1],[-1,0],[0,-1]]
        for i in range(m):
            for j in range(n):
                if mat[i][j] == 0:
                    q.append([i,j])
                    
                else:
                    mat[i][j] = -1
                    
        
        while q:
            size = len(q)
            for _ in range(size):
                r,c = q.popleft()                
                for dr,dc in dirs:
                    nr = dr+r
                    nc = dc+c                    
                    if nr>=0 and nr<m and nc>=0 and nc<n and mat[nr][nc]==-1:
                        mat[nr][nc] = dist
                        q.append([nr,nc])
            dist+=1
            
        return mat
                    